### PR TITLE
PyNumber and PyFunction to wrap python objects

### PR DESCRIPTION
**Pull request by isuruf**
Example ipynb : https://gist.github.com/isuruf/82081aed4cb25ef98799

### CODE REVIEW

**Comment by certik** [comment]
Looks good to me. Should some of the stuff from `pywrapper.h/cpp` go into symengine itself? I assume it will have to be replicated for Ruby and Julia as well.

**Comment by isuruf** [comment]
Some of the functionality of `PyFunctionClass` would have to be replicated, but let's worry about that when Ruby and Julia wrappers are at that stage.

**Comment by certik** [comment]
Cool, resolve the conflicts and things should start working.

**Comment by isuruf** [comment]
Merged with master. Tests are passing now

File: symengine/tests/test_functions.py [hunk 1]
```diff
@@ -88,12 +88,31 @@ def test_Subs():
...
-    sympy.var("n, m, theta, phi")
+    n, m, theta, phi = sympy.symbols("n, m, theta, phi")
     r = sympy.Ynm(n, m, theta, phi)
     s = Integer(2)*r
     assert isinstance(s, Mul)
     assert isinstance(s.args[1]._sympy_(), sympy.Ynm)
 
+    x = symbols("x")
+    e = x + sympy.loggamma(x)
+    assert str(e) == "x + loggamma(x)"
```
Comment (certik) [inline symengine/tests/test_functions.py]: You should also check here that the expression is a symengine expression, not sympy expression.

**Comment by certik** [comment]
Also, since you implemented Sage support, we should add tests for Sage.

File: symengine/lib/symengine/pywrapper.h [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    RCP<const Basic> (*diff_)(PyObject*, RCP<const Basic>);
+    PyObject *one, *zero, *minus_one;
+public:
+    PyModule(PyObject* (*)(const RCP<const Basic> x), RCP<const Basic> (*)(PyObject*),
+             RCP<const Number> (*)(PyObject*, long), RCP<const Basic> (*)(PyObject*, RCP<const Basic>));
+    ~PyModule();
+    PyObject* get_zero() const { return zero; }
+    PyObject* get_one() const { return one; }
+    PyObject* get_minus_one() const { return minus_one; }
+};
```
Comment (certik) [inline symengine/lib/symengine/pywrapper.h]: Add documentation for `PyModule` about what it does and how it is meant to be used. If I understand it correctly, the higher level code must provide Python functions `from_py_`, `eval_` and `diff_`, that are stored in this `PyModule`. They are then called from C++.

File: symengine/lib/symengine/pywrapper.h [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    virtual RCP<const Number> rpow(const Number &other) const;
+
+    //! Differentiation w.r.t Symbol `x`
+    virtual RCP<const Basic> diff(const RCP<const Symbol> &x) const;
+    virtual RCP<const Number> eval(long bits) const;
+    virtual std::string __str__() const;
+    virtual int compare(const Basic &o) const;
+    virtual bool __eq__(const Basic &o) const;
+    virtual std::size_t __hash__() const;
+};
```
Comment (certik) [inline symengine/lib/symengine/pywrapper.h]: Add documentation. It seems that arithmetic operations are implemented by calling functions like `PyNumber_Divide`, but I wasn't able to find definitions of those. Are these Python C / API?

File: symengine/lib/symengine/pywrapper.h [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    PyFunctionClass(PyObject *pyobject, std::string name, const RCP<const PyModule> &pymodule);
+    PyObject* get_py_object() const { return pyobject_; }
+    RCP<const PyModule> get_py_module() const { return pymodule_; }
+    std::string get_name() const { return name_; }
+
+    PyObject* call(const vec_basic &vec) const;
+    bool __eq__(const PyFunctionClass &x) const;
+    int compare(const PyFunctionClass &x) const;
+    std::size_t hash() const;
+};
```
Comment (certik) [inline symengine/lib/symengine/pywrapper.h]: Add documentation.

File: symengine/lib/symengine/pywrapper.h [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+    PyObject *get_py_object() const;
+    RCP<const PyFunctionClass> get_pyfunction_class() const;
+    virtual RCP<const Basic> create(const vec_basic &x) const;
+    virtual RCP<const Number> eval(long bits) const;
+    virtual RCP<const Basic> diff(const RCP<const Symbol> &x) const;
+    virtual int compare(const Basic &o) const;
+    virtual bool __eq__(const Basic &o) const;
+    virtual std::size_t __hash__() const;
+};
```
Comment (certik) [inline symengine/lib/symengine/pywrapper.h]: Add documentation.

File: symengine/lib/symengine/pywrapper.h (hunk 3 above)
Comment (isuruf) [inline symengine/lib/symengine/pywrapper.h]: Yes, they are Python C/API functions.

File: symengine/tests/test_functions.py (hunk 1 above)
Comment (certik) [inline symengine/tests/test_functions.py]: Also add more tests to exercise all the methods like +, -, / that you implemented.

**Comment by certik** [comment]
Also add upper level documentation into the C++ module about how it works. My understanding is that you subclass SymEngine's C++ classes, and then from within them, you call Python functions to implement the virtual methods. Then in the Python wrapper, you implement those Python functions in Cython, and pass them to this C++ subclass. This should be clearly documented, so that other people can quickly understand what is going on.

The only other comment I have is that I generally don't like to use the Python C/API and especially the reference counting directly, but rather using Cython. But I think the current solution works and is robust. And once it is merged, we can think if there is a way to move most of the stuff to Cython, and only leave bare C++ subclasses that dispatch directly to Cython.

**Comment by isuruf** [comment]
Added comments and more tests including sage

**Comment by certik** [comment]
I think it looks good now. +1 to merge if tests pass.

**Comment by isuruf** [comment]
Made a change to the tests.

**Comment by certik** [comment]
That's fine. Merging. Thanks for the PR!

**Pull request merged**